src/GafferSceneUI/CropWindowTool.cpp

```diff
@@ -451,13 +555,17 @@ void CropWindowTool::plugDirtied( const Gaffer::Plug *plug )
 {
 	if( plug == activePlug() )
 	{
-		m_overlay->setVisible( activePlug()->getValue() );
-		m_needCropWindowPlugSearch = m_overlayDirty = true;
```

Comment:
Dirtying taken care of in scene plug search.

Reply:
That makes the dirty propagation more complex and more fragile. If I happen to call `findCropWindowPlug()` before I call `findScenePlug()`, the necessary updates won't happen and I'll end up with the wrong cropWindowPlug.

Reply:
Back to our chat the other day that we need a better way to manage all this state...

Reply:
But as `findScenePlug()` always dirties the crop plug, even if you call `findCropPlug()` first, when you call `findScenePlug()` ill will re-instate the crop dirtyness so you shouldn't end up with the crop plug being wrong?

Reply:
1) Starting state : neither is dirty
2) Active plug is dirtied : now only m_needScenePlugSearch is dirty
3) I call `findCropWindowPlug()` : it returns the wrong plug.

> Back to our chat the other day that we need a better way to manage all this state...

A good starting point is to not rely on side effects from the find functions, other than finding the thing they said they world. If we keep dirty/find as two distinct steps, then we're closer to being able to separate them back out. Yes, if we had two different "encapsulated lazy state trackers", the crop window one would probably dirty itself when the scene plug one was dirtied. But _at the point of dirtying_, not at the point the lazy search is done.

Reply:
Revised version in https://github.com/GafferHQ/gaffer/pull/3582/commits/decc9e63f72b3e323a087c669fdded47906e371a as per our offline chats. A note for the record:

 - We ultimately want to refactor the state tracking inside the tool as its way too fragile as is. 
 - EditScopes are going to require a further refactor, so we'll tackle this then rather than now.